### PR TITLE
[AMORO-2662]After the token expires,the optimizer repeatedly commits tasks that have been reset to ams

### LINKED ISSUE #2662: [Bug]: After the token expires,the optimizer repeatedly commits tasks that have been reset to ams
### What happened?

After the token expires,,the optimizer repeatedly commits tasks that have been reset to ams, so that new tasks cannot be pulled.

### Affects Versions

master

### What engines are you seeing the problem on?

AMS

### How to reproduce

1. Start an optimizer with only one thread
2. optimizer `polledTask` and `ackedTask`
3. token expire and reset task
4. `completeTask`
5. Then you will find that the optimizer performs `completeTask` repeatedly.

### Relevant log output

```shell
2024-03-11 18:20:00,388 ERROR com.netease.arctic.optimizer.common.AbstractOptimizerOperator [] [flink-optimizer-executor-1] - Call ams got an error and will try again later
com.netease.arctic.ams.api.ArcticException: java.lang.IllegalStateException: Task not scheduled yet, taskId:OptimizingTaskId(processId:1710151650912, taskId:30)
    at com.netease.arctic.ams.api.OptimizingService$completeTask_result$completeTask_resultStandardScheme.read(OptimizingService.java:4937) ~[MuiseOptimizer.jar:?]
    at com.netease.arctic.ams.api.OptimizingService$completeTask_result$completeTask_resultStandardScheme.read(OptimizingService.java:4923) ~[MuiseOptimizer.jar:?]
    at com.netease.arctic.ams.api.OptimizingService$completeTask_result.read(OptimizingService.java:4873) ~[MuiseOptimizer.jar:?]
    at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:88) ~[MuiseOptimizer.jar:?]
    at com.netease.arctic.ams.api.OptimizingService$Client.recv_completeTask(OptimizingService.java:176) ~[MuiseOptimizer.jar:?]
    at com.netease.arctic.ams.api.OptimizingService$Client.completeTask(OptimizingService.java:162) ~[MuiseOptimizer.jar:?]
    at sun.reflect.GeneratedMethodAccessor58.invoke(Unknown Source) ~[?:?]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_121]
    at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_121]
    at com.netease.arctic.ams.api.client.ThriftClientPool.lambda$iface$1(ThriftClientPool.java:261) ~[MuiseOptimizer.jar:?]
    at com.sun.proxy.$Proxy37.completeTask(Unknown Source) ~[?:?]
    at com.netease.arctic.optimizer.common.OptimizerExecutor.lambda$completeTask$2(OptimizerExecutor.java:142) ~[MuiseOptimizer.jar:?]
    at com.netease.arctic.optimizer.common.AbstractOptimizerOperator.callAuthenticatedAms(AbstractOptimizerOperator.java:95) ~[MuiseOptimizer.jar:?]
    at com.netease.arctic.optimizer.common.OptimizerExecutor.completeTask(OptimizerExecutor.java:140) ~[MuiseOptimizer.jar:?]
    at com.netease.arctic.optimizer.common.OptimizerExecutor.start(OptimizerExecutor.java:53) ~[MuiseOptimizer.jar:?]
    at com.netease.arctic.optimizer.flink.FlinkExecutor.lambda$open$0(FlinkExecutor.java:41) ~[MuiseOptimizer.jar:?]
    at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by rfyu**
## Why are the changes needed?

Close #2662

## Brief change log

- Added an IllegalTaskException

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: ams/server/src/main/java/com/netease/arctic/server/exception/IllegalTaskException.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netease.arctic.server.exception;
+
+public class IllegalTaskException extends ArcticRuntimeException {
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/exception/IllegalTaskException.java:21]: Perhaps we don't need to declare an exception class for every scenario; a more general exception might suffice for some common situations.

So I recommend to rename the `DuplicateRuntimeException` to `TaskRuntimeException`. All the general errors encountered during the task processing can all be wrapped using it. We may just need to describe the detail in the error message.

HDYT? @rfyu

**Review by zhoujinsong — COMMENTED** [review]
@rfyu I left some comments, PTAL.

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/TaskRuntime.java [hunk 2]
```diff
@@ -299,7 +300,7 @@ public String toString() {
 
   private void validThread(OptimizerThread thread) {
     if (token == null) {
-      throw new IllegalStateException("Task not scheduled yet, taskId:" + taskId);
+      throw new IllegalTaskException("Task has been reset or not yet scheduled, taskId:" + taskId);
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/TaskRuntime.java:303]: Good job, the error description is more proper now.

But we may improve the error message in line 305 too, like:
```
The optimizer thread does not match, (xxx) in the task, (xxx) in the request.
```

**Review by zhoujinsong — COMMENTED** [review]
I left a small suggestion. PTAL.

File: ams/server/src/main/java/com/netease/arctic/server/exception/IllegalTaskException.java (hunk 1 above)
Comment (rfyu) [reply ams/server/src/main/java/com/netease/arctic/server/exception/IllegalTaskException.java:21]: Agreed.I will make changes.

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/TaskRuntime.java [hunk 3]
```diff
@@ -299,10 +299,17 @@ public String toString() {
 
   private void validThread(OptimizerThread thread) {
     if (token == null) {
-      throw new IllegalStateException("Task not scheduled yet, taskId:" + taskId);
+      throw new TaskRuntimeException("Task has been reset or not yet scheduled, taskId:%s", taskId);
     }
-    if (!thread.getToken().equals(getToken()) || thread.getThreadId() != threadId) {
-      throw new DuplicateRuntimeException("Task already acked by optimizer thread + " + thread);
+    if (!thread.getToken().equals(getToken())) {
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/TaskRuntime.java:304]: I believe displaying both the token and thread ID in the error logs would facilitate problem localization more conveniently. Therefore, I suggest making some changes like:
```
    if (!thread.getToken().equals(getToken()) || thread.getThreadId() != threadId) {
      throw new TaskRuntimeException(
          "The optimizer threads do not match, the thread in the task is OptimizerThread(token=%s, threadId=%s), and the thread in the request is OptimizerThread(token=%s, threadId=%s).",
          getToken(), threadId, thread.getToken(), thread.getThreadId());
    }
```

**Review by zhoujinsong — APPROVED** [review]
LGTM.
Thanks a lot for your contribution!

**Pull request merged**